Update ext/standard parameter names (WIP)

nikic opened the pull request:
Just putting this up to avoid duplicate work ... will take a while to go through all of it.

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -408,23 +408,23 @@ function dns_check_record(string $hostname, string $type = "MX"): bool {}
+ * @param array $additional_records
  */
-function dns_get_record(string $hostname, int $type = DNS_ANY, &$authns = null, &$addtl = null, bool $raw = false): array|false {}
+function dns_get_record(string $hostname, int $type = DNS_ANY, &$authoritative_name_servers = null, &$additional_records = null, bool $raw = false): array|false {}
```
These names are excessively long, but I'm not familiar enough with this API to suggest a good concise name.

kocsismate commented on ext/standard/basic_functions.stub.php:
```diff
@@ -529,53 +529,53 @@ function get_html_translation_table(int $table = HTML_SPECIALCHARS, int $quote_s
+function explode(string $separator, string $string, int $limit = PHP_INT_MAX): array {}
 
-function implode(string|array $glue, ?array $pieces = null): string {}
+function implode(string|array $separator, ?array $array = null): string {}
```
adieu `$glue` and `$pieces`, you won't be missed :D

kocsismate commented on ext/standard/basic_functions.stub.php:
```diff
@@ -600,50 +600,50 @@ function str_starts_with(string $haystack, string $needle): bool {}
+function lcfirst(string $string): string {}
 
-function ucwords(string $str, string $delimiters = " \t\r\n\f\v"): string {}
+function ucwords(string $string, string $delimiters = " \t\r\n\f\v"): string {}
```
Is there any difference between a "delimiter" and a "separator"?

nikic replied:
The docs say:
> The optional delimiters contains the word separator characters. 
So, I'd say, no, there is no difference :)

I've also adjusted the csv APIs accordingly (comma *separated* values, not comma delimited values!)

kocsismate commented on ext/standard/basic_functions.stub.php:
```diff
@@ -600,50 +600,50 @@ function str_starts_with(string $haystack, string $needle): bool {}
+function ucwords(string $string, string $delimiters = " \t\r\n\f\v"): string {}
 
-function strtr(string $str, string|array $from, ?string $to = null): string {}
+function strtr(string $string, string|array $from, ?string $to = null): string {}
```
Slightly unrelated: I saw that the manual has a very detailed description about the two versions of `strtr()` (https://www.php.net/manual/en/function.strtr.php), and this is the first time when I realized that the default value fix for PHP 8.0 has a side-effect that we can't have an entirely accurate name for the 2nd parameter :/

nikic replied:
Something to discuss, but I think what we should do with these overloaded signatures is to keep listing them as multiple signatures in the manual and define a policy that overloaded functions are excluded from our parameter name stability policy -- and that we reserve the right to make invoking them with named params a hard error in the future (once we have "positional only" arguments). We just can't support these kinds of signatures properly.

kocsismate commented on ext/standard/basic_functions.stub.php:
```diff
@@ -600,50 +600,50 @@ function str_starts_with(string $haystack, string $needle): bool {}
+function stripslashes(string $string): string {}
 
-/** @param int $replace_count */
-function str_replace(array|string $search, array|string $replace, string|array $subject, &$replace_count = null): string|array {}
```
Bob had a very good suggestion IMO to rename these to `$from`, `$to`, `$subject`. Sometimes I'm also wondering if `$search` is what we search for or where we search, so I'd also prefer these names. :)

nikic replied:
strtr() uses $from, $to terminology, so there's some precedent there...

Something we also need to consider is how this will interact with pcre APIs. Here we have `$search, $replace, $subject`. In preg_replace, we have `$pattern, $replace, $subject`. In preg_replace_callback, we have `$pattern, $callback, $subject`.

kocsismate commented on ext/standard/basic_functions.stub.php:
```diff
@@ -600,50 +600,50 @@ function str_starts_with(string $haystack, string $needle): bool {}
+function str_ireplace(array|string $search, array|string $replace, string|array $subject, &$count = null): string|array {}
 
-function hebrev(string $str, int $max_chars_per_line = 0): string {}
+function hebrev(string $string, int $max_chars_per_line = 0): string {}
```
nit: in other places `character` is not abbreviated. I'm ok with `$max_chars_per_line` though, it already has very long a name.

kocsismate commented on ext/standard/basic_functions.stub.php:
```diff
@@ -652,41 +652,41 @@ function strip_tags(string $str, array|string|null $allowable_tags = null): stri
+function strpbrk(string $string, string $characters): string|false {}
 
-function substr_compare(string $main_str, string $str, int $offset, ?int $length = null, bool $case_insensitivity = false): int {}
+function substr_compare(string $main_string, string $string, int $offset, ?int $length = null, bool $case_insensitive = false): int {}
```
Bob's suggestions were `$haystack` and `$needle`, and I also pretty much love these names.

kocsismate replied:
And that's all I could check today.

nikic replied:
Makes sense, I went with it.

kocsismate reviewed (commented):
Other ideas:
- `array_push()`: I think we could use `$array` instead of `$stack` (update: I now see that you did this change in a few places)
- `array_walk*()`: `$array` instead of `$input`, `$args` instead of `$argument`
- `array_filter()`: `$flags` instead of `$use_keys`
- `move_uploaded_file()`:  *maybe* "from" and "to" could be used (either also including `path` or not: `$from`/`$from_path`), but I like the source - destination terms too
- `headers_sent()`: `$filename`

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -113,33 +113,33 @@ function in_array(mixed $needle, array $haystack, bool $strict = false): bool {}
 function compact($var_name, ...$var_names): array {}
 
-function array_fill(int $start_key, int $count, mixed $value): array {}
+function array_fill(int $start, int $length, mixed $value): array {}
```
I think my thought process here was to mirror array_slice terminology, but I'm not sure how much sense that really makes. I changed it back to `$count`.

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -113,33 +113,33 @@ function in_array(mixed $needle, array $haystack, bool $strict = false): bool {}
-function array_fill(int $start_key, int $count, mixed $value): array {}
+function array_fill(int $start, int $length, mixed $value): array {}
 
 function array_fill_keys(array $keys, mixed $value): array {}
```
I think keeping `$keys` is a bit nicer here, due to the key(s) + value combination in this API.

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -113,33 +113,33 @@ function in_array(mixed $needle, array $haystack, bool $strict = false): bool {}
+function array_unshift(array &$array, mixed ...$values): int {}
 
-function array_splice(array &$array, int $offset, ?int $length = null, mixed $replacements = []): array {}
+function array_splice(array &$array, int $offset, ?int $length = null, mixed $replacement = []): array {}
```
After thinking about this a bit, I've normalized things to use `$offset` everywhere (in ext/standard -- I'll update ext/mbstring if we agree on this). Based on current use, `$offset` is more common (by 2-3 times) and a bit more flexible in usage (the strrpos issue mentioned). It's also already used in file-system related APIs, where it seems more natural (especially as in some APIs like fseek it can be an offset from different "whence").

kocsismate replied:
Yeah, seems a sensible idea!

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -244,32 +244,32 @@ function array_key_exists($key, array $array): bool {}
 
 function constant(string $name): mixed {}
 
 function ip2long(string $ip_address): int|false {}
```
Heh. I'm not actually sure I prefer it, but I went with it to commemorate the occasion ;)

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -244,32 +244,32 @@ function array_key_exists($key, array $array): bool {}
+function getenv(?string $name = null, bool $local_only = false): string|array|false {}
 
 #ifdef HAVE_PUTENV
 function putenv(string $setting): bool {}
```
`man putenv` uses the inspired name ... `$string`. I went with `$assignment`, as I like it better than `$setting` at least.

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -244,32 +244,32 @@ function array_key_exists($key, array $array): bool {}
-/** @param int $optind */
-function getopt(string $options, array $longopts = [], &$optind = null): array|false {}
+/** @param int $rest_index */
+function getopt(string $options, array $long_options = [], &$rest_index = null): array|false {}
```
Done

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -285,23 +285,23 @@ function time_sleep_until(float $timestamp): bool {}
-function get_cfg_var(string $option_name): string|array|false {}
+function get_cfg_var(string $name): string|array|false {}
 
 function error_log(string $message, int $message_type = 0, ?string $destination = null, ?string $extra_headers = null): bool {}
```
Done

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -85,20 +85,20 @@ function uasort(array &$array, callable $callback): bool {}
-function key(array|object $arg): int|string|null {}
+function key(array|object $array): int|string|null {}
 
 function min(mixed $arg, mixed ...$args): mixed {}
```
Done

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -312,28 +312,28 @@ function php_strip_whitespace(string $filename): string {}
 function highlight_string(string $string, bool $return = false): string|bool {}
 
-function ini_get(string $varname): string|false {}
+function ini_get(string $name): string|false {}
```
Hm, personally this doesn't seem ambiguous to me.

kocsismate replied:
Then the `$name` seems to be ok as it works well in the other functions like `ini_set`.

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -344,16 +344,16 @@ function getservbyport(int $port, string $protocol): string|false {}
-function unregister_tick_function(callable $function): void {}
+function unregister_tick_function(callable $callback): void {}
 
 function is_uploaded_file(string $path): bool {}
```
I was thinking about this as well. My understanding of the current usage here is `$filename` is a path to a file, `$directory` is a path to a directory and `$path` is a path to a file or a directory. Following that rule, this should indeed be `$filename`. But I don't know if the rule is right :)

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -434,22 +434,22 @@ function net_get_interfaces(): array|false {}
 /* md5.c */
 
-function md5(string $str, bool $raw_output = false): string {}
+function md5(string $string, bool $binary = false): string {}
```
I'm using `$binary` here to be consistent with what we did in ext/hash and ext/openssl.

nikic replied:
But of course we can change this. Actually, the names of `bool` parameters are the most important, as they are most likely to be used with named parameters -- one could say that most other names don't matter at all...

I think the `as_` prefix from above works quite well in a lot of these cases, i.e `microtime(as_float: true)`, `md5($str, as_binary: true)`. The `return_` one also works `microtime(return_float: true)`, `md5($str, return_binary: true)`, but is a good bit longer. The `use_` prefix doesn't fit well for the `microtime(use_float: true)` case imho.

kocsismate replied:
I agree, `use_` is the least good option. I like `as_` and `return_` too, but I agree, `as_` is better. All in all, I couldn't choose between `as_` and no prefix.

nikic replied:
For some case like `hrtime($as_number)`, not having the prefix might be confusing as `hrtime($number)` looks like it accepts a number argument, like many other functions do. For other cases like `$md5($string, $binary)` it seems unambiguous without a prefix, but may still make sense to add for consistency. I don't know :)

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -489,10 +489,10 @@ function inet_pton(string $ip_address): string|false {}
 
 /* {{{ head.c */
-function header(string $string, bool $replace = true, int $http_response_code = 0): void {}
+function header(string $header, bool $replace = true, int $http_response_code = 0): void {}
```
Done. I don't think it's ambiguous here, as header clearly deals with HTTP.

kocsismate replied:
Agreed!

nikic commented:
>     * `array_push()`: I think we could use `$array` instead of `$stack` (update: I now see that you did this change in a few places)

Done! This was indeed just an oversight.

>     * `array_walk*()`: `$array` instead of `$input`, `$args` instead of `$argument`

I've renamed `$input` to `$array` and `$argument` to `$arg`. However, I think it should stay in the singular, as you can only pass through a single argument.

>     * `array_filter()`: `$flags` instead of `$use_keys`

Hm... it looks like the values here don't form a bitmaks. I've changed it to `$mode`, does that seem reasonable?
 
>     * `move_uploaded_file()`:  _maybe_ "from" and "to" could be used (either also including `path` or not: `$from`/`$from_path`), but I like the source - destination terms too

I'm not sure on this one. We seem to be using "old"/"new" for renaming operations (see also `rename`) and "source"/"dest" for copy operations (see also `copy`, `stream_copy_to_stream`).

>     * `headers_sent()`: `$filename`

Done.

kocsismate commented on ext/standard/basic_functions.stub.php:
```diff
@@ -1152,7 +1152,7 @@ function password_algos(): array {}
  * @return resource|false
  */
-function proc_open(array|string $cmd, array $descriptorspec, &$pipes, ?string $cwd = null, ?array $env = null, ?array $other_options = null) {}
+function proc_open(array|string $command, array $descriptor_spec, &$pipes, ?string $cwd = null, ?array $env = null, ?array $options = null) {}
```
maybe `$env_vars` (`$env` seems to be ok too)

nikic replied:
Done

kocsismate commented on ext/standard/basic_functions.stub.php:
```diff
@@ -1198,7 +1198,7 @@ function soundex(string $string): string {}
 /* streamsfuncs.c */
 
-function stream_select(?array &$read, ?array &$write, ?array &$except, ?int $tv_sec, int $tv_usec = 0): int|false {}
+function stream_select(?array &$read, ?array &$write, ?array &$except, ?int $seconds, int $microsectons = 0): int|false {}
```
```suggestion
function stream_select(?array &$read, ?array &$write, ?array &$except, ?int $seconds, int $microseconds = 0): int|false {}
```

nikic replied:
Ooops :)

kocsismate commented on ext/standard/basic_functions.stub.php:
```diff
@@ -1487,7 +1488,7 @@ function version_compare(string $version1, string $version2, ?string $operator =
 
 #ifdef PHP_WIN32
-function sapi_windows_cp_set(int $cp): bool {}
+function sapi_windows_cp_set(int $codepage): bool {}
```
This seems to be written as two words (just like codepoint above).  I guess they could stay as-is, if you prefer writing them as a single word.

nikic replied:
We are using `$codepoint` as one word everywhere though (49 times).

kocsismate replied:
ok, let's go with `$codepoint` then

kocsismate reviewed (commented):
Oh gosh, I finally reached the bottom of the stub file. :D (although, this time I only vaguely checked the non-changed parts).

- `closedir`: should `$dir_handle` become a `$stream`?
- `password_needs_rehash`: `$hash` to `$password`?

kocsismate commented:
> I've renamed $input to $array and $argument to $arg. However, I think it should stay in the singular, as you can only pass through a single argument.

the plural was just a typo ^^

kocsismate commented:
> Hm... it looks like the values here don't form a bitmaks. I've changed it to $mode, does that seem reasonable?

Yes, absolutely!

'm not sure on this one. We seem to be using "old"/"new" for renaming operations (see also rename) and "source"/"dest" for copy operations (see also copy, stream_copy_to_stream).

I'm on board :)

nikic commented:
>     * `closedir`: should `$dir_handle` become a `$stream`?

Not sure on this one: You are right that closedir() accepts a stream resource, but it's a special kind (PHP_STREAM_FLAG_IS_DIR) created by opendir(). Passing something like fopen() to it will throw. Is it worthwhile to distinguish this?

>     * `password_needs_rehash`: `$hash` to `$password`?

This function accepts a hash from password_hash(), not a plain password.

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -923,7 +923,7 @@ function lchgrp(string $filename, string|int $group): bool {}
 
 #if HAVE_UTIME
-function touch(string $filename, ?int $time = null, ?int $atime = null): bool {}
+function touch(string $filename, ?int $mtime = null, ?int $atime = null): bool {}
```
I've kept these terms because we have filemtime() and fileatime() and also return these keys from stat/etc. Not really opposed to more explicit names though.

kocsismate replied:
ok, this reasoning seems sensible for me as well, so let's keep it as-is.

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -941,39 +941,39 @@ function realpath_cache_size(): int {}
  * @return resource|false
  */
-function fsockopen(string $hostname, int $port = -1, &$errno = null, &$errstr = null, ?float $timeout = null) {}
+function fsockopen(string $hostname, int $port = -1, &$errno = null, &$error_message = null, ?float $timeout = null) {}
```
I wanted to retain the `$errno` term to make it clear that this is a C errno, but I agree that the juxtaposition of `$errno` and `$error_message` is rather jarring, so I've changed this to `$error_code`.

kocsismate replied:
Yeah, thanks! My main problem was this indeed.

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -1107,7 +1107,7 @@ function decoct(int $number): string {}
-function base_convert(string $number, int $frombase, int $tobase): string {}
+function base_convert(string $number, int $from_base, int $to_base): string {}
 
 function number_format(float $number, int $decimals = 0, ?string $decimal_point = "." , ?string $thousands_separator = ","): string {}
```
Done

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -1107,7 +1107,7 @@ function decoct(int $number): string {}
 function dechex(int $number): string {}
 
-function base_convert(string $number, int $frombase, int $tobase): string {}
+function base_convert(string $number, int $from_base, int $to_base): string {}
```
I've updated this to hopefully be in line with bcmath/gmp now.

nikic commented on ext/standard/basic_functions.stub.php:
```diff
@@ -800,45 +800,45 @@ function fread($handle, int $length): string|false {}
+function rename(string $old_name, string $new_name, $context = null): bool {}
 
 /** @param resource|null $context */
 function copy(string $source, string $dest, $context = null): bool {}
```
I've pushed a separate commit (https://github.com/php/php-src/pull/6214/commits/8c7578e9dcaeaf06c452a278c76dc81af0f9df1a currently) that normalizes to $from/$to across various functions. Does it look reasonable? I've not added filename/path/stream suffixes as it seamed obvious from context.

kocsismate replied:
Nice! I didn't really like the old/new terminology, and using only from/to still seems clear.

kocsismate commented:
> Not sure on this one: You are right that closedir() accepts a stream resource, but it's a special kind (PHP_STREAM_FLAG_IS_DIR) created by opendir(). Passing something like fopen() to it will throw. Is it worthwhile to distinguish this?

Yes, you're right then, I wasn't really aware of the distinction. Will `$dir_handle` still work well when we convert this resource to object? Ah, I've just checked it, and since `Directory::close()` is the implementation alias of `closedir`, I guess we could convert the resource to `Directory`, no? Thus, I'd go with `$directory` if all this is really possible.

> This function accepts a hash from password_hash(), not a plain password.

Ah, yes, I didn't realize at first that it's for distinguishing hashed password from plain password. This is something I also support very much, so +1!

nikic commented:
> > Not sure on this one: You are right that closedir() accepts a stream resource, but it's a special kind (PHP_STREAM_FLAG_IS_DIR) created by opendir(). Passing something like fopen() to it will throw. Is it worthwhile to distinguish this?
> 
> Yes, you're right then, I wasn't really aware of the distinction. Will `$dir_handle` still work well when we convert this resource to object? Ah, I've just checked it, and since `Directory::close()` is the implementation alias of `closedir`, I guess we could convert the resource to `Directory`, no? Thus, I'd go with `$directory` if all this is really possible.

Hard to say. Depends on how the conversion is done. If we use the same underlying implementation, directories might also be an instance of `Stream` rather than a separate object.

I think `$directory` is a fine name here -- my only concern is that it does not give an obvious distinction between a directory string and a directory resource.

kocsismate commented:
@nikic The PR seems good for me now! :) Somewhere you asked a question about the `as_` for bool params, but I can't find it now. I think we could still fine tune these kind of param names a bit later. So all in all, I think this PR is good for landing for RC1, and we could do smaller amendments and "are we sure we're good with it"-kind PRs if necessary next week.

nikic commented:
@kocsismate Yeah, I think these are the open questions that are left:
 * `$as_` prefix.
 * `$dir_handle`
 * `str_replace` naming (let's solve this one while handling ext/pcre).

We'll have to do some fine tuning later.

kocsismate reviewed (commented):
Ok, I'm leaving two more small ideas here:
- `parse_url`: first param could be `$uri` because it accepts URIs. It doesn't match with the function name though, so this change is not a must-have. P.S. ext/ldap is about to use `$uri`.
- `is_callable()`: `$callable_name` could be `$name` or `$callback_name` I think.

nikic commented:
Another things we'll have to check is that manual references to `argument #n ($xxx)` in error messages are correct. I'm sure some of them have been broken.
